# Day 1

Calorie Counting
---

The jungle must be too overgrown and difficult to navigate in vehicles or access from the air; the Elves' expedition traditionally goes on foot. As your boats approach land, the Elves begin taking inventory of their supplies. One important consideration is food - in particular, the number of Calories each Elf is carrying (your puzzle input).

The Elves take turns writing down the number of Calories contained by the various meals, snacks, rations, etc. that they've brought with them, one item per line. Each Elf separates their own inventory from the previous Elf's inventory (if any) by a blank line.

For example, suppose the Elves finish writing their items' Calories and end up with the following list:

```
1000
2000
3000

4000

5000
6000

7000
8000
9000

10000
```

This list represents the Calories of the food carried by five Elves:

- The first Elf is carrying food with 1000, 2000, and 3000 Calories, a total of 6000 Calories.
- The second Elf is carrying one food item with 4000 Calories.
- The third Elf is carrying food with 5000 and 6000 Calories, a total of 11000 Calories.
- The fourth Elf is carrying food with 7000, 8000, and 9000 Calories, a total of 24000 Calories.
- The fifth Elf is carrying one food item with 10000 Calories.

In case the Elves get hungry and need extra snacks, they need to know which Elf to ask: they'd like to know how many Calories are being carried by the Elf carrying the most Calories. In the example above, this is 24000 (carried by the fourth Elf).

Find the Elf carrying the most Calories. **How many total Calories is that Elf carrying?**

To begin, get your puzzle input.

Answer

### Read file and retrieve calories

In [1]:
calories_file = "elf_calories.txt"

In [4]:
with open(calories_file, "r") as f:
    file_contents = f.read()

In [5]:
file_contents

'15560\n3906\n7076\n11980\n11508\n\n6558\n2256\n7294\n6566\n2686\n2566\n1724\n4811\n5427\n4278\n3756\n\n3761\n5599\n13187\n12558\n7425\n9269\n\n59658\n\n34175\n9643\n\n7664\n10224\n5541\n1378\n1784\n9347\n7692\n\n3786\n4260\n4416\n6345\n6750\n6415\n5467\n4898\n2017\n4486\n1198\n1960\n\n5658\n6168\n8088\n7749\n5890\n1980\n6014\n1136\n3138\n6706\n7041\n\n9197\n6997\n12644\n12609\n7630\n12928\n\n5278\n1423\n4822\n5148\n6387\n1280\n6431\n3847\n5886\n2714\n3725\n6331\n\n7537\n6965\n10079\n8904\n8639\n10679\n8964\n\n4426\n8844\n4298\n2872\n10619\n5095\n11321\n\n7983\n1883\n8608\n5411\n8970\n7877\n5185\n3210\n8719\n\n3726\n3043\n3978\n6669\n1915\n5948\n2128\n5272\n4685\n5797\n2780\n\n4455\n10134\n10707\n2796\n5472\n8624\n\n13936\n11059\n4848\n1048\n8927\n1803\n\n11965\n3983\n8834\n12932\n2111\n10667\n\n5514\n19118\n\n1826\n2484\n4686\n3811\n1299\n3538\n1627\n4180\n3148\n2443\n6232\n4090\n3027\n4364\n\n1020\n3031\n9049\n2732\n3458\n8787\n9492\n9701\n\n5906\n7234\n7874\n5787\n1655\n5841\n2750\n

### Create dictionary with calories split by elves
`elf_dict`

In [7]:
split_elves = file_contents.split("\n\n")
split_elves[:10]

['15560\n3906\n7076\n11980\n11508',
 '6558\n2256\n7294\n6566\n2686\n2566\n1724\n4811\n5427\n4278\n3756',
 '3761\n5599\n13187\n12558\n7425\n9269',
 '59658',
 '34175\n9643',
 '7664\n10224\n5541\n1378\n1784\n9347\n7692',
 '3786\n4260\n4416\n6345\n6750\n6415\n5467\n4898\n2017\n4486\n1198\n1960',
 '5658\n6168\n8088\n7749\n5890\n1980\n6014\n1136\n3138\n6706\n7041',
 '9197\n6997\n12644\n12609\n7630\n12928',
 '5278\n1423\n4822\n5148\n6387\n1280\n6431\n3847\n5886\n2714\n3725\n6331']

In [20]:
elf_dict = dict()
for elf in range(len(split_elves)):
    if "\n" in split_elves[elf]:
        elf_dict[elf] = [int(i) for i in split_elves[elf].split("\n")]
    else:
        elf_dict[elf] = [int(split_elves[elf])]

In [31]:
elf_dict

{0: [15560, 3906, 7076, 11980, 11508],
 1: [6558, 2256, 7294, 6566, 2686, 2566, 1724, 4811, 5427, 4278, 3756],
 2: [3761, 5599, 13187, 12558, 7425, 9269],
 3: [59658],
 4: [34175, 9643],
 5: [7664, 10224, 5541, 1378, 1784, 9347, 7692],
 6: [3786, 4260, 4416, 6345, 6750, 6415, 5467, 4898, 2017, 4486, 1198, 1960],
 7: [5658, 6168, 8088, 7749, 5890, 1980, 6014, 1136, 3138, 6706, 7041],
 8: [9197, 6997, 12644, 12609, 7630, 12928],
 9: [5278, 1423, 4822, 5148, 6387, 1280, 6431, 3847, 5886, 2714, 3725, 6331],
 10: [7537, 6965, 10079, 8904, 8639, 10679, 8964],
 11: [4426, 8844, 4298, 2872, 10619, 5095, 11321],
 12: [7983, 1883, 8608, 5411, 8970, 7877, 5185, 3210, 8719],
 13: [3726, 3043, 3978, 6669, 1915, 5948, 2128, 5272, 4685, 5797, 2780],
 14: [4455, 10134, 10707, 2796, 5472, 8624],
 15: [13936, 11059, 4848, 1048, 8927, 1803],
 16: [11965, 3983, 8834, 12932, 2111, 10667],
 17: [5514, 19118],
 18: [1826,
  2484,
  4686,
  3811,
  1299,
  3538,
  1627,
  4180,
  3148,
  2443,
  6232,
  4090,

### Obtain the total calories collected by elf

In [32]:
total_calories = {key: sum(value) for key, value in elf_dict.items()}

In [33]:
total_calories

{0: 50030,
 1: 47922,
 2: 51799,
 3: 59658,
 4: 43818,
 5: 43630,
 6: 51998,
 7: 59568,
 8: 62005,
 9: 53272,
 10: 61767,
 11: 47475,
 12: 57846,
 13: 45941,
 14: 42188,
 15: 41621,
 16: 50492,
 17: 24632,
 18: 46755,
 19: 47270,
 20: 51773,
 21: 43199,
 22: 46236,
 23: 54530,
 24: 35778,
 25: 54786,
 26: 52210,
 27: 44973,
 28: 47412,
 29: 51130,
 30: 50158,
 31: 54018,
 32: 60613,
 33: 63252,
 34: 50663,
 35: 56904,
 36: 62565,
 37: 52473,
 38: 46419,
 39: 15842,
 40: 57990,
 41: 62289,
 42: 48749,
 43: 57308,
 44: 18033,
 45: 46833,
 46: 45353,
 47: 35239,
 48: 59676,
 49: 69528,
 50: 24572,
 51: 55793,
 52: 38865,
 53: 44652,
 54: 51753,
 55: 36228,
 56: 49129,
 57: 36152,
 58: 45906,
 59: 57662,
 60: 37983,
 61: 62924,
 62: 52327,
 63: 43542,
 64: 26864,
 65: 56934,
 66: 44860,
 67: 42014,
 68: 39147,
 69: 58192,
 70: 55441,
 71: 3308,
 72: 49464,
 73: 44776,
 74: 51430,
 75: 58111,
 76: 40795,
 77: 40361,
 78: 43933,
 79: 61750,
 80: 41600,
 81: 54490,
 82: 31677,
 83: 56853,
 84

### Most calorific

In [42]:
all_calories = total_calories.values()
all_calories

dict_values([50030, 47922, 51799, 59658, 43818, 43630, 51998, 59568, 62005, 53272, 61767, 47475, 57846, 45941, 42188, 41621, 50492, 24632, 46755, 47270, 51773, 43199, 46236, 54530, 35778, 54786, 52210, 44973, 47412, 51130, 50158, 54018, 60613, 63252, 50663, 56904, 62565, 52473, 46419, 15842, 57990, 62289, 48749, 57308, 18033, 46833, 45353, 35239, 59676, 69528, 24572, 55793, 38865, 44652, 51753, 36228, 49129, 36152, 45906, 57662, 37983, 62924, 52327, 43542, 26864, 56934, 44860, 42014, 39147, 58192, 55441, 3308, 49464, 44776, 51430, 58111, 40795, 40361, 43933, 61750, 41600, 54490, 31677, 56853, 57703, 7578, 44530, 18247, 29620, 67746, 51944, 42929, 51956, 35075, 55097, 18474, 50075, 47252, 45302, 53973, 46549, 47599, 14472, 26990, 47486, 53317, 46228, 55463, 45080, 12986, 43156, 38853, 54902, 51806, 47006, 45279, 49770, 48452, 46798, 37947, 32701, 49925, 23081, 55713, 36621, 39104, 55986, 54970, 34847, 19897, 60110, 47832, 8994, 49710, 47469, 57402, 33890, 44469, 31683, 16083, 66006, 529

In [43]:
maximum_cal = max(all_calories)
maximum_cal

69528

### Find the Elf carrying the most Calories. **How many total Calories is that Elf carrying?**

In [38]:
total_calories.items()

dict_items([(0, 50030), (1, 47922), (2, 51799), (3, 59658), (4, 43818), (5, 43630), (6, 51998), (7, 59568), (8, 62005), (9, 53272), (10, 61767), (11, 47475), (12, 57846), (13, 45941), (14, 42188), (15, 41621), (16, 50492), (17, 24632), (18, 46755), (19, 47270), (20, 51773), (21, 43199), (22, 46236), (23, 54530), (24, 35778), (25, 54786), (26, 52210), (27, 44973), (28, 47412), (29, 51130), (30, 50158), (31, 54018), (32, 60613), (33, 63252), (34, 50663), (35, 56904), (36, 62565), (37, 52473), (38, 46419), (39, 15842), (40, 57990), (41, 62289), (42, 48749), (43, 57308), (44, 18033), (45, 46833), (46, 45353), (47, 35239), (48, 59676), (49, 69528), (50, 24572), (51, 55793), (52, 38865), (53, 44652), (54, 51753), (55, 36228), (56, 49129), (57, 36152), (58, 45906), (59, 57662), (60, 37983), (61, 62924), (62, 52327), (63, 43542), (64, 26864), (65, 56934), (66, 44860), (67, 42014), (68, 39147), (69, 58192), (70, 55441), (71, 3308), (72, 49464), (73, 44776), (74, 51430), (75, 58111), (76, 40795)

In [41]:
[(key, value) for key, value in total_calories.items() if value == maximum_cal]

[(49, 69528)]

## Part 2

Your puzzle answer was 69528.

The first half of this puzzle is complete! It provides one gold star: *

--- Part Two ---
By the time you calculate the answer to the Elves' question, they've already realized that the Elf carrying the most Calories of food might eventually run out of snacks.

To avoid this unacceptable situation, the Elves would instead like to know the total Calories carried by the top three Elves carrying the most Calories. That way, even if one of those Elves runs out of snacks, they still have two backups.

In the example above, the top three Elves are the fourth Elf (with 24000 Calories), then the third Elf (with 11000 Calories), then the fifth Elf (with 10000 Calories). The sum of the Calories carried by these three elves is 45000.

Find the top three Elves carrying the most Calories. How many Calories are those Elves carrying in total?

Answer:


In [63]:
sorted_calories = sorted(list(all_calories), reverse=True)

In [65]:
sorted_calories[:3]

[69528, 68878, 67746]

In [66]:
sum(sorted_calories[:3])

206152